File: frontend/src/services/types/material.ts
import { PaginationResult } from './common';

// 器材创建模式
export interface MaterialCreate {
  material_code: string;
  material_name: string;
  material_specification?: string;
  material_desc?: string;
  material_wdh?: string;
  safety_stock?: number;
  equipment_id?: number;
}

// 器材更新模式
export interface MaterialUpdate {
  material_code?: string;
  material_name?: string;
  material_specification?: string;
  material_desc?: string;
  material_wdh?: string;
  safety_stock?: number;
  major_id?: number;
  equipment_id?: number;
}

// 器材响应模式
export interface MaterialResponse {
  id: number;
  material_code: string;
  material_name: string;
  material_specification?: string;
  material_desc?: string;
  material_wdh?: string;
  safety_stock?: number;
  material_query_code?: string;
  major_id?: number;
  major_name?: string;
  equipment_id?: number;
  equipment_name?: string;
  creator?: string;
  create_time?: string;
  update_time?: string;
}

// 器材查询参数
export interface MaterialQueryParams {
  material_code?: string;
  material_name?: string;
  material_specification?: string;
  material_desc?: string;
  material_wdh?: string;
  safety_stock?: number;
  major_id?: number;
  major_name?: string;
  equipment_id?: number;
  equipment_name?: string;
  creator?: string;
  search?: string;
  sort_field?: string;
  sort_asc?: boolean;
}

// 器材分页参数
export interface MaterialPaginationParams {
  page: number;
  page_size: number;
  material_code?: string;
  material_name?: string;
  material_specification?: string;
  material_desc?: string;
  material_wdh?: string;
  safety_stock?: number;
  major_id?: number;
  major_name?: string;
  equipment_id?: number;
  equipment_name?: string;
  creator?: string;
  search?: string;
  sort_field?: string;
  sort_asc?: boolean;
}

// 器材分页结果
export interface MaterialPaginationResult extends PaginationResult<MaterialResponse> {
  // 继承PaginationResult的所有属性，data属性已包含在泛型中
}

// 器材列表响应
export interface MaterialListResponse {
  data: MaterialResponse[];
  total: number;
}

// 器材统计信息
export interface MaterialStatistics {
  total_count: number;
  major_count: Record<string, number>;
  equipment_count: Record<string, number>;
}

// 批量删除器材
export interface BatchMaterialDelete {
  ids: number[];
}

// 批量导入结果
export interface MaterialBatchImportResult {
  success_count: number;
  error_count: number;
  errors: string[];
}

// 器材表单数据（用于创建和编辑）
export interface MaterialFormData {
  material_code: string;
  material_name: string;
  material_specification?: string;
  material_desc?: string;
  material_wdh?: string;
  safety_stock?: number;
  equipment_id?: number;
}

// 器材搜索参数（简化版，用于组件搜索）
export interface MaterialSearchParams {
  search?: string;
  material_code?: string;
  material_name?: string;
  equipment_id?: number;
  major_id?: number;
}

// 器材选择项（用于下拉选择）
export interface MaterialOption {
  value: number;
  label: string;
  material_code: string;
  material_name: string;
}

// 装备选项（用于专业下的装备选择）
export interface EquipmentOption {
  id: number;
  display_name: string;
}

// 装备选项响应
export interface EquipmentOptionsResponse {
  data: EquipmentOption[];
}

// 专业选项
export interface MajorOption {
  id: number;
  major_name: string;
}

// 专业选项集合响应模型
export interface MajorOptionsResponse {
  data: MajorOption[];
  total_count: number;
}

// 器材定位查询参数
export interface MaterialLocateParams {
  material_code?: string;  // 器材编码(可选,为空则返回第1页)
  page_size?: number;       // 每页大小,用于计算页码
}

// 器材定位结果
export interface MaterialLocateResult {
  found: boolean;                    // 是否找到匹配器材
  material: MaterialResponse | null; // 匹配的器材详情
  position: number | null;           // 器材在ID升序列表中的位置(从1开始)
  target_page: number | null;        // 器材所在页码
  page_size: number;                 // 用于计算页码的每页大小
  total_before: number | null;       // 该器材之前的器材数量(ID更小的器材数)
  suggestions: MaterialResponse[];   // 保留字段(始终为空列表)
}

File: backend/schemas/base/material.py
from pydantic import BaseModel, Field
from typing import Optional, List
from datetime import datetime
from schemas.common.base import PaginationResult

# 器材创建模式
class MaterialCreate(BaseModel):
    material_code: str
    material_name: str
    material_specification: Optional[str] = None
    material_desc: Optional[str] = None
    material_wdh: Optional[str] = None
    safety_stock: Optional[int] = None
    material_query_code: Optional[str] = None  # 用户可选的查询码字段
    equipment_id: Optional[int] = None

# 器材更新模式
class MaterialUpdate(BaseModel):
    material_code: Optional[str] = None
    material_name: Optional[str] = None
    material_specification: Optional[str] = None
    material_desc: Optional[str] = None
    material_wdh: Optional[str] = None
    safety_stock: Optional[int] = None
    material_query_code: Optional[str] = None  # 用户可选的查询码字段
    major_id: Optional[int] = None
    equipment_id: Optional[int] = None

# 器材响应模式
class MaterialResponse(BaseModel):
    id: int
    material_code: str
    material_name: str
    material_specification: Optional[str]
    material_desc: Optional[str]
    material_wdh: Optional[str]
    safety_stock: Optional[int]
    material_query_code: Optional[str]
    major_id: Optional[int]
    major_name: Optional[str]
    equipment_id: Optional[int]
    equipment_name: Optional[str]
    creator: Optional[str]
    create_time: Optional[datetime]
    update_time: Optional[datetime]

    class Config:
        from_attributes = True

# 器材查询参数
class MaterialQueryParams(BaseModel):
    material_code: Optional[str] = None
    material_name: Optional[str] = None
    material_specification: Optional[str] = None
    material_desc: Optional[str] = None
    material_wdh: Optional[str] = None
    safety_stock: Optional[int] = None
    major_id: Optional[int] = None
    major_name: Optional[str] = None
    equipment_id: Optional[int] = None
    equipment_name: Optional[str] = None
    creator: Optional[str] = None
    search: Optional[str] = None
    sort_field: Optional[str] = "id"
    sort_asc: Optional[bool] = True

# 器材分页参数
class MaterialPaginationParams(BaseModel):
    page: int = Field(1, ge=1, description="页码，从1开始")
    page_size: int = Field(10, ge=1, le=1000, description="每页记录数，最大100")
    material_code: Optional[str] = None
    material_name: Optional[str] = None
    material_specification: Optional[str] = None
    material_desc: Optional[str] = None
    material_wdh: Optional[str] = None
    safety_stock: Optional[int] = None
    major_id: Optional[int] = None
    major_name: Optional[str] = None
    equipment_id: Optional[int] = None
    equipment_name: Optional[str] = None
    creator: Optional[str] = None
    search: Optional[str] = None
    sort_field: Optional[str] = "id"
    sort_asc: Optional[bool] = True

# 器材分页结果
class MaterialPaginationResult(PaginationResult):
    data: List[MaterialResponse]

# 器材列表响应
class MaterialListResponse(BaseModel):
    data: List[MaterialResponse]
    total: int

# 器材统计信息
class MaterialStatistics(BaseModel):
    total_count: int
    major_count: dict
    equipment_count: dict

# 批量删除器材
class BatchMaterialDelete(BaseModel):
    ids: List[int]

# 批量导入结果
class MaterialBatchImportResult(BaseModel):
    success_count: int
    error_count: int
    errors: List[str]

# 准专业选项响应模型
class MajorOption(BaseModel):
    id: int
    major_name: str

class MajorOptionsResponse(BaseModel):
    data: List[MajorOption]

# 装备选项响应模型
class EquipmentOption(BaseModel):
    id: int
    display_name: str

class EquipmentOptionsResponse(BaseModel):
    data: List[EquipmentOption]
    total: int

# 器材定位查询参数
class MaterialLocateParams(BaseModel):
    material_code: str = Field(..., description="器材编码(仅支持精确匹配)")
    page_size: int = Field(10, ge=1, le=100, description="每页大小,用于计算页码")

# 器材定位结果
class MaterialLocateResult(BaseModel):
    found: bool = Field(..., description="是否找到匹配器材")
    material: Optional[MaterialResponse] = Field(None, description="匹配的器材详情")
    position: Optional[int] = Field(None, description="器材在ID升序列表中的位置(从1开始)")
    target_page: Optional[int] = Field(None, description="器材所在页码")
    page_size: int = Field(10, description="用于计算页码的每页大小")
    total_before: Optional[int] = Field(None, description="该器材之前的器材数量(ID更小的器材数)")
    suggestions: List[MaterialResponse] = Field(default_factory=list, description="保留字段(始终为空列表)")